### PR TITLE
DataStore Clear API

### PR DESCRIPTION
**Pull request by Ashish-Nanda**
*Issue #:*  #160

*Description of changes:*

This commit adds the code to implement the `Clear` functionality on both iOS and Android for the DataStore category of Amplify Flutter.
Includes unit tests.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt [hunk 1]
```diff
@@ -198,6 +204,22 @@ class AmplifyDataStorePlugin : FlutterPlugin, MethodCallHandler {
...
+    fun onClear(flutterResult: Result) {
+        val plugin = Amplify.DataStore.getPlugin("awsDataStorePlugin") as AWSDataStorePlugin
+
+        plugin.clear(
+                {
+                    LOG.debug("Successfully cleared the store")
+                    handler.post { flutterResult.success(null) }
+                },
+                {
+                    LOG.debug("Failed to clear store with error: " + it.localizedMessage)
```
Comment (Amplifiyer) [inline packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt:217]: Probably log.error?

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt [hunk 2]
```diff
@@ -63,14 +67,17 @@ class AmplifyDataStorePlugin : FlutterPlugin, MethodCallHandler {
     override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
         var data: HashMap<String, Any> = HashMap()
         try {
-            data = checkArguments(call.arguments) as HashMap<String, Any>
+            if(call.arguments!= null) {
```
Comment (Amplifiyer) [inline packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt:70]: nit: space needed between arguments and !

File: packages/amplify_datastore/android/src/test/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePluginTest.kt [hunk 3]
```diff
@@ -197,6 +198,44 @@ class AmplifyDataStorePluginTest {
             )
     }
 
+    @Test
+    fun test_Clear_Success_Result() {
+        var dataStoreException = DataStoreException("AmplifyException", DataStoreException.REPORT_BUG_TO_AWS_SUGGESTION)
```
Comment (Amplifiyer) [inline packages/amplify_datastore/android/src/test/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePluginTest.kt:203]: Probably not needed here.

File: packages/amplify_datastore/example/ios/amplify_datastore_unit_tests/DataStorePluginUnitTests.swift [hunk 4]
```diff
@@ -334,4 +334,51 @@ class DataStorePluginUnitTests: XCTestCase {
...
+        }
+    }
+        let dataStoreBridge: MockDataStoreBridge = MockDataStoreBridge()
+        pluginUnderTest = SwiftAmplifyDataStorePlugin(bridge: dataStoreBridge, flutterModelRegistration: flutterModelSchemaRegistration)
+        
+        pluginUnderTest.onClear(
+            flutterResult: {(result) in
+                XCTAssertNil(result)
+            })
+    }
```
Comment (Amplifiyer) [inline packages/amplify_datastore/example/ios/amplify_datastore_unit_tests/DataStorePluginUnitTests.swift:354]: Indentation seems off. You can run the Xcode's default formatter for indentation. We will later invest in swiftlint etc.

File: packages/amplify_datastore/example/lib/main.dart [hunk 5]
```diff
@@ -158,26 +167,46 @@ class _MyAppState extends State<MyApp> {
...
-                    'All Blogs \n$_blogs\n\n' +
-                    'First post from list of all posts\n$_firstPostFromResult\n\n' +
-                    'Posts >= 4 rating\n$_posts4rating\n\n' +
-                    'Posts between 1 and 4 rating\n$_posts1To4Rating\n\n' +
-                    'Posts with rating 2 or 5\n$_posts2Or5Rating\n\n' +
-                    'Posts without rating 2 or 5\n$_allPostsWithoutRating2Or5\n\n' +
-                    'Post with date equals\n$_postWithCreatedDate\n\n' +
-                    'Post with Id not equals\n$_postWithIdNotEquals\n\n')),
-          )),
+        ),
```
Comment (Amplifiyer) [inline packages/amplify_datastore/example/lib/main.dart:208]: If we don't intend to keep these, we should remove the calls and vars declared earlier.

**Review by Amplifiyer — COMMENTED** [review]
Looks good, left some minor comments.

File: packages/amplify_core_plugin_interface/lib/amplify_datastore_category.dart [hunk 6]
```diff
@@ -59,6 +59,10 @@ class DataStoreCategory {
         : throw (errorMsg);
   }
 
+  Future<void> clear() {
+    return plugins.length == 1 ? plugins[0].clear() : throw (errorMsg);
+  }
+
```
Comment (jamesonwilliams) [inline packages/amplify_core_plugin_interface/lib/amplify_datastore_category.dart:65]:
```dart
  Future<void> clear() {
    return _selectPlugin((it) => it.clear());
  }

  Future<R> _selectPlugin<R>(FutureOr<R> onValue(DataStoreCategoryInterface plugin)) {
    return plugins.length == 1 ?
      Future.value(plugins[0]).then(onValue) :
      throw("DataStore plugin not added correctly");
  }
```

File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt [hunk 7]
```diff
@@ -198,6 +204,22 @@ class AmplifyDataStorePlugin : FlutterPlugin, MethodCallHandler {
         )
     }
 
+    @VisibleForTesting
+    fun onClear(flutterResult: Result) {
+        val plugin = Amplify.DataStore.getPlugin("awsDataStorePlugin") as AWSDataStorePlugin
+
+        plugin.clear(
+                {
+                    LOG.debug("Successfully cleared the store")
```
Comment (jamesonwilliams) [inline packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt:213]: Might be significant enough for an `info` level?

**Review by jamesonwilliams — APPROVED** [review]
Looks good!

**Review by Ashish-Nanda — COMMENTED** [review]
Addressed comments, and added additional improvements and tests

**Review by Amplifiyer — APPROVED** [review]
:shipit:

File: packages/amplify_core_plugin_interface/lib/amplify_datastore_category.dart (hunk 6 above)
Comment (jamesonwilliams) [reply packages/amplify_core_plugin_interface/lib/amplify_datastore_category.dart:65]: Sounds great!

**Review by jamesonwilliams — APPROVED** [review]
To these eyes, it looks good! 🚢

**Pull request merged**